### PR TITLE
feat: optionally remove self from staked validators cache

### PR DESCRIPTION
**Pull request by ksn6**
#### Problem
`StakedValidatorsCache` always has the locally running validator's TPU vote socket in its cache entry.

#### Summary of Changes
Add the ability for the locally running validator to opt out of adding its own TPU vote socket to the cache entry. This avoids the validator from sending a vote to itself.

### CODE REVIEW

File: core/src/staked_validators_cache.rs [hunk 1]
```diff
@@ -39,6 +39,9 @@ pub struct StakedValidatorsCache {
 
     /// Protocol
     protocol: Protocol,
+
+    /// Whether to include the running validator's socket address in the list of staked validators
```
Comment (wen-coding) [inline core/src/staked_validators_cache.rs]:
```suggestion
    /// Whether to include the running validator's socket address
```

File: core/src/staked_validators_cache.rs [hunk 2]
```diff
@@ -89,7 +94,15 @@ impl StakedValidatorsCache {
...
             .iter()
-            .filter(|(_, stake)| **stake > 0)
+            .filter(|(pubkey, stake)| {
+                let positive_stake = **stake > 0;
+
+                if self.include_self {
+                    positive_stake
+                } else {
+                    positive_stake && pubkey != &&cluster_info.id()
+                }
```
Comment (wen-coding) [inline core/src/staked_validators_cache.rs]: nit: we can also do
```
let positive_stake = **stake > 0;
let not_self = pubkey != &cluster_info.id();
is_positive && (self.include_self || not_self)
```

I think there is a slight chance cluster_info.id() doesn't match pubkey if you are doing set-identity, but it's not end of the world if I sometimes send a few packets to myself if I do set-identity, so I think this is good enough.

File: core/src/staked_validators_cache.rs [hunk 3]
```diff
@@ -378,12 +390,14 @@ mod tests {
...
-            .bank_forks();
+        let (bank_forks, cluster_info) =
+            StakedValidatorsCacheHarness::new(&genesis_config, Keypair::new())
+                .with_vote_accounts(slot_num, keypair_map, vahm, protocol)
+                .bank_forks();
 
         // Create our staked validators cache
-        let mut svc = StakedValidatorsCache::new(bank_forks, protocol, Duration::from_secs(5), 5);
+        let mut svc =
+            StakedValidatorsCache::new(bank_forks, protocol, Duration::from_secs(5), 5, true);
```
Comment (wen-coding) [inline core/src/staked_validators_cache.rs:397]: nit: this is a test, I guess we don't need 5 seconds here? Making it 1 second can make the test run faster?

File: core/src/staked_validators_cache.rs [hunk 4]
```diff
@@ -532,4 +549,62 @@ mod tests {
         let (_, refreshed) = svc.get_staked_validators_by_slot(2 * slot_num, &cluster_info, now);
         assert!(refreshed);
     }
+
+    #[test_case(325_000_000_u64, 1_usize, 10_usize, 123_u64, Protocol::UDP)]
+    #[test_case(325_000_000_u64, 3_usize, 10_usize, 123_u64, Protocol::QUIC)]
+    #[test_case(325_000_000_u64, 10_usize, 10_usize, 123_u64, Protocol::UDP)]
+    #[test_case(325_000_000_u64, 10_usize, 10_usize, 123_u64, Protocol::QUIC)]
+    #[test_case(325_000_000_u64, 50_usize, 60_usize, 123_u64, Protocol::UDP)]
+    fn test_exclude_self_from_cache(
```
Comment (wen-coding) [inline core/src/staked_validators_cache.rs]: nit: This new test is a function test, I guess we don't need to test all combinations, just testing one should be more than enough.

File: core/src/staked_validators_cache.rs (hunk 1 above)
Comment (ksn6) [reply core/src/staked_validators_cache.rs]: Would rather leave the more detailed comment. Updated it for clarity.

File: core/src/staked_validators_cache.rs (hunk 2 above)
Comment (ksn6) [reply core/src/staked_validators_cache.rs]: I originally had these three lines haha. I think it's a bit more clear the current way by expanding the code out more explicitly.

File: core/src/staked_validators_cache.rs (hunk 3 above)
Comment (ksn6) [reply core/src/staked_validators_cache.rs:397]: All tests in `staked_validators_cache.rs` use simulated time, so e.g., even increasing 5 seconds to 60 seconds wouldn't make a difference; most of these tests take at most a few milliseconds to run.

File: core/src/staked_validators_cache.rs (hunk 4 above)
Comment (ksn6) [reply core/src/staked_validators_cache.rs]: I'm testing that this works for n = 1, 3, 10, and 50 validators - probably a good idea to test for edge cases, just in case.

EDIT: these tests collectively take ~0.12 seconds to run; not too much overhead.

File: core/src/staked_validators_cache.rs (hunk 1 above)
Comment (wen-coding) [reply core/src/staked_validators_cache.rs]: Sure, my comment was mostly about this line being a bit long.

File: core/src/staked_validators_cache.rs (hunk 2 above)
Comment (wen-coding) [reply core/src/staked_validators_cache.rs]: The current logic is very short, so it's probably fine either way. But I think your original three lines are much clearer, because "let not_self = pubkey != &cluster_info.id();" clearly explains your intention on doing the test, so if some day a newbie comes in and reads the code, this is totally self explanatory.
The current code, on the other hand, needs the code reader to dig in and see why the logic is different when self.include_self is false, which IMHO is a bit less self explanatory.

File: core/src/staked_validators_cache.rs (hunk 4 above)
Comment (wen-coding) [reply core/src/staked_validators_cache.rs]: I don't feel strongly about this one case, probably fine either way.

That said, tests are the supplementary documents of code. If you list this many combinations, I would wonder whether they make a difference in your logic. (That is, why don't we test more combinations, why these numbers.)
I think testing 1 and multiple validators totally makes sense, say 1 and 10. But I don't think the code path changes a lot on 3 vs 10 validators. If I need to add more tests, it probably makes more sense to add some contact_info with no tpu_vote port than to test 3 vs 10 validators here. Because that adds more test coverage. Just my 2c.

File: core/src/staked_validators_cache.rs (hunk 2 above)
Comment (ksn6) [reply core/src/staked_validators_cache.rs]: Sure - no strong opinions on this from me.

File: core/src/staked_validators_cache.rs (hunk 4 above)
Comment (ksn6) [reply core/src/staked_validators_cache.rs]: Cool - keeping 1 and 10, but changing between UDP and QUIC for now.

**Review by wen-coding — APPROVED** [review]
